services/m3dbnode/main/main.go

```diff
@@ -23,114 +23,383 @@ package main
-	tchannelNodeAddrArg    = flag.String("nodetchanneladdr", "0.0.0.0:9003", "Node TChannel server address")
+	configFile = flag.String("f", "", "configuration file")
 )
 
 func main() {
 	flag.Parse()
 
-	if *httpClusterAddrArg == "" ||
-		*tchannelClusterAddrArg == "" ||
-		*httpNodeAddrArg == "" ||
-		*tchannelNodeAddrArg == "" {
+	if len(*configFile) == 0 {
 		flag.Usage()
 		os.Exit(1)
 	}
 
-	id := *idArg
-	httpClusterAddr := *httpClusterAddrArg
-	tchannelClusterAddr := *tchannelClusterAddrArg
-	httpNodeAddr := *httpNodeAddrArg
-	tchannelNodeAddr := *tchannelNodeAddrArg
+	var cfg config.Configuration
+	if err := xconfig.LoadFile(&cfg, *configFile); err != nil {
+		fmt.Fprintf(os.Stderr, "unable to load %s: %v", *configFile, err)
+	}
+
+	logger, err := cfg.Logging.BuildLogger()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "unable to create logger: %v", err)
+	}
 
-	storageOpts := storage.NewOptions()
-	log := storageOpts.InstrumentOptions().Logger()
+	debug.SetGCPercent(cfg.GCPercentage)
 
-	namespaces, err := server.DefaultNamespaces()
+	scope, _, err := cfg.Metrics.NewRootScope()
 	if err != nil {
-		log.Fatalf("could not create default namespaces: %v", err)
+		logger.Fatalf("could not connect to metrics: %v", err)
 	}
 
-	storageOpts = storageOpts.SetNamespaceInitializer(
-		namespace.NewStaticInitializer(namespaces))
-	topoInit, err := server.DefaultTopologyInitializer(id, tchannelNodeAddr)
+	opts := storage.NewOptions()
+
+	iopts := opts.InstrumentOptions().
+		SetLogger(logger).
+		SetMetricsScope(scope).
+		SetMetricsSamplingRate(cfg.Metrics.SampleRate())
+	opts = opts.SetInstrumentOptions(iopts)
+
+	buildReporter := instrument.NewBuildReporter(iopts)
+	if err := buildReporter.Start(); err != nil {
+		logger.Fatalf("unable to start build reporter: %v", err)
+	}
+	defer buildReporter.Close()
+
+	configSvcClientOpts := cfg.ConfigService.NewOptions().
+		SetInstrumentOptions(
+			instrument.NewOptions().
+				SetLogger(logger).
+				SetMetricsScope(scope))
+	configSvcClient, err := etcdclient.NewConfigServiceClient(configSvcClientOpts)
 	if err != nil {
-		log.Fatalf("could not create topology initializer: %v", err)
+		logger.Fatalf("could not create m3cluster client: %v", err)
 	}
 
-	if id == "" {
-		id, err = os.Hostname()
-		if err != nil {
-			log.Fatalf("could not get hostname: %v", err)
-		}
+	dynamicOpts := namespace.NewDynamicOptions().
+		SetInstrumentOptions(iopts).
+		SetConfigServiceClient(configSvcClient).
+		SetNamespaceRegistryKey(kvconfig.NamespacesKey)
+	nsInit := namespace.NewDynamicInitializer(dynamicOpts)
+
+	opts = opts.
+		SetTickInterval(cfg.TickInterval).
+		SetNamespaceInitializer(nsInit)
+
+	newFileMode, err := cfg.Filesystem.ParseNewFileMode()
+	if err != nil {
+		log.Fatalf("could not parse new file mode: %v", err)
+	}
+
+	newDirectoryMode, err := cfg.Filesystem.ParseNewFileMode()
+	if err != nil {
+		log.Fatalf("could not parse new directory mode: %v", err)
+	}
+
+	fsopts := fs.NewOptions().
```

Comment:
The main would get simplified a lot if we could just get a `storage.Options` from the top level config struct. Could you make `Options()` a method on the config structs and do this stuff there?

Reply:
As discussed, this is a big change that we should probably defer considering there are other bigger issues at hand.  We can always revisit this when there is some more time.